Support for customizing AllowedIPRange of NodePorts

ahmedwaleedmalik opened the pull request:
Signed-off-by: Waleed Malik <ahmedwaleedmalik@gmail.com>

### What this PR does / why we need it
Support for customizing AllowedIPRange of NodePorts.

### Which issue(s) this PR fixes
<!--
Use one of the GitHub's keywords to link connected Issues/PRs.
Keyword list: https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
Fixes #3924 

### Special notes for your reviewer
<!-- Remove if not needed -->

### Release note
<!--
Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just leave "NONE".
-->
```release-note
Support for customizing AllowedIPRange of NodePorts 
```

maciaszczykm commented on src/app/wizard/step/provider-settings/provider/extended/openstack/default/component.ts:
```diff
@@ -173,6 +185,8 @@ export class OpenstackProviderExtendedDefaultCredentialsComponent
       case Controls.SubnetID:
         return this._canLoadSubnet() ? '' : 'Please enter your credentials and network first.';
+      default:
+        return null;
```
This method should return string, change the signature or this line to return empty string.

maciaszczykm replied:
Is this resolved?

maciaszczykm commented on src/app/wizard/step/provider-settings/provider/extended/openstack/default/template.html:
```diff
@@ -65,4 +65,17 @@
   </km-combobox>
+
+  <mat-form-field fxFlex>
+    <mat-label>Allowed IP range for NodePorts</mat-label>
```
Use title case.

maciaszczykm replied:
It is not resolved.

ahmedwaleedmalik replied:
I fixed this locally and marked it as resolved, since I was about to push my changes. But then found out a few other issues :)

maciaszczykm commented on src/app/wizard/step/provider-settings/provider/extended/gcp/component.ts:
```diff
@@ -151,13 +163,18 @@ export class GCPProviderExtendedComponent extends BaseFormValidator implements O
       case Controls.SubNetwork:
         return this._hasRequiredCredentials() ? '' : 'Please enter your credentials first.';
     }
+    return null;
```
Move it to default inside switch and look on my comment above regarding typings.

ahmedwaleedmalik commented:
@maciaszczykm  Initially, I added `NodePortsAllowedIPRange` field in all the supported cloud providers and their individual forms with https://github.com/kubermatic/dashboard/pull/4198/commits/2f68a33bfd5ac2112d2beed3a024bf0019811263. 

IMO it adds a lot of redundancy so I switched to https://github.com/kubermatic/dashboard/pull/4198/commits/67906ee8aa7df72e8e9ed05bb51e464805df06cb. That is introducing the field at the root level for extended provider settings instead of for each individual cloud provider  https://github.com/ahmedwaleedmalik/dashboard/blob/allowed-ip-ranges/src/app/wizard/step/provider-settings/provider/extended/template.html#L47

WDYT of this approach? Presets won't have any effect on this field for now.

floreks commented on src/app/shared/constants/constants.ts:
```diff
@@ -0,0 +1,15 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export const CIDR_ALLOW_ALL = '0.0.0.0/0';
```
Export this to a static class similar to how it's done in `src/app/shared/validators/validators.ts`.

floreks replied:
Why was it marked as is resolved without any change or comment?

ahmedwaleedmalik replied:
I fixed this locally and marked it as resolved since I was about to push my changes. But then found out a few other issues :)

ahmedwaleedmalik replied:
Logic has been modified for this, PTAL.

floreks commented on src/app/wizard/step/provider-settings/provider/extended/component.ts:
```diff
@@ -47,13 +50,21 @@ export class ProviderExtendedComponent extends BaseFormValidator implements OnIn
   ngOnInit(): void {
     this.form = this._builder.group({
       [Controls.ProviderExtended]: this._builder.control(''),
+      [Controls.NodePortsAllowedIPRange]: this._builder.control(CIDR_ALLOW_ALL, [
```
Is it really a required property?

ahmedwaleedmalik replied:
It shouldn't be empty if this is enabled

floreks replied:
I've asked if it is a required property meaning that a value has to be provided in order to proceed further.

ahmedwaleedmalik replied:
Logic has been modified for this, PTAL.

floreks commented on src/app/wizard/step/provider-settings/provider/extended/component.ts:
```diff
@@ -62,5 +73,22 @@ export class ProviderExtendedComponent extends BaseFormValidator implements OnIn
+  }
+
+  private _onNodePortsAllowedIPRangeChange(value: string): void {
+    if (this.Providers.AWS) {
```
Did you actually test it for other providers and checked console logs for errors?

ahmedwaleedmalik replied:
I just modified this and was looking for thoughts on https://github.com/kubermatic/dashboard/pull/4198#issuecomment-1046653060

It's still WIP

maciaszczykm replied:
If it is work in progress mark it as draft or WIP. It was mentioned on Slack too.

floreks commented on src/app/wizard/step/provider-settings/provider/extended/component.ts:
```diff
@@ -47,13 +50,21 @@ export class ProviderExtendedComponent extends BaseFormValidator implements OnIn
   readonly Controls = Controls;
 
+  // This is only supported on AWS, Azure, GCP, and Openstack
+  supportsAllowedIPRangeOverride =
+    this.Providers.AWS || this.Providers.AZURE || this.Providers.GCP || this.Providers.OPENSTACK;
```
How is it supposed to work? How does it check for the currently selected provider? It basically seems to be an equivalent of `true || true || true || true`.

maciaszczykm commented on src/app/shared/constants/constants.ts:
```diff
@@ -0,0 +1,15 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export const CIDR_ALLOW_ALL = '0.0.0.0/0';
```
Is this a new file and directory? Is this used I more than one place?

ahmedwaleedmalik replied:
This has been removed.

floreks commented on src/app/wizard/step/provider-settings/provider/extended/template.html:
```diff
@@ -43,4 +43,19 @@
+           type="text"
+           autocomplete="off">
+    <mat-hint>IP address range that is allowed to access NodePorts </mat-hint>
+    <mat-error *ngIf="form.get(Controls.NodePortsAllowedIPRange).hasError( 'pattern')">
```
s/`( 'pattern')`/`('pattern')`

maciaszczykm commented on src/app/wizard/step/provider-settings/provider/extended/component.ts:
```diff
@@ -62,5 +73,22 @@ export class ProviderExtendedComponent extends BaseFormValidator implements OnIn
+  }
+
+  private _onNodePortsAllowedIPRangeChange(value: string): void {
+    if (this.Providers.AWS) {
```
This should be switch? Also, check @floreks comment.

maciaszczykm commented on src/app/wizard/step/provider-settings/provider/extended/component.ts:
```diff
@@ -47,20 +50,69 @@ export class ProviderExtendedComponent extends BaseFormValidator implements OnIn
+
+  private _onNodePortsAllowedIPRangeChange(value: string): void {
+    switch (this.provider) {
+      case this.Providers.AWS:
```
this.Providers is imported there so we can use it in the template, inside the component use the enum directly.

floreks commented:
Closing in favour of https://github.com/kubermatic/dashboard/pull/4314

/close
